Hmm, this toolchain did not appear to work. Building the current `main` of Hummingbird using the current `main` of swift-nio (by putting swift-nio in edit mode) produces the same errors using the downloaded Ubuntu 20.04 toolchain as we see with the current 6.0 toolchain:

```
/hummingbird/Sources/HummingbirdTLS/TLSChannel.swift:40:44: error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
38 |     @inlinable
39 |     public func setup(channel: Channel, logger: Logger) -> EventLoopFuture<Value> {
40 |         return channel.pipeline.addHandler(NIOSSLServerHandler(context: self.sslContext)).flatMap {
   |                                            `- error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
41 |             self.baseChannel.setup(channel: channel, logger: logger)
42 |         }

/hummingbird/.build/checkouts/swift-nio-ssl/Sources/NIOSSL/NIOSSLHandler.swift:791:1: note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 789 | 
 790 | @available(*, unavailable)
 791 | extension NIOSSLHandler: Sendable {}
     | `- note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 792 | 
 793 | extension NIOSSLHandler {
/hummingbird/Sources/HummingbirdTLS/TLSChannel.swift:40:44: error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
38 |     @inlinable
39 |     public func setup(channel: Channel, logger: Logger) -> EventLoopFuture<Value> {
40 |         return channel.pipeline.addHandler(NIOSSLServerHandler(context: self.sslContext)).flatMap {
   |                                            `- error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
41 |             self.baseChannel.setup(channel: channel, logger: logger)
42 |         }

/hummingbird/.build/checkouts/swift-nio-ssl/Sources/NIOSSL/NIOSSLHandler.swift:791:1: note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 789 | 
 790 | @available(*, unavailable)
 791 | extension NIOSSLHandler: Sendable {}
     | `- note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 792 | 
 793 | extension NIOSSLHandler {
```

This is the [`addHandler` method](https://github.com/apple/swift-nio/blob/6d30ec4738389e875f1a22345463468d4d581063/Sources/NIOCore/ChannelPipeline.swift#L170-L175) in question, with the `@preconcurrency` annotation on it.

--- a/lib/Sema/TypeCheckAvailability.cpp
+++ b/lib/Sema/TypeCheckAvailability.cpp
@@ -3524,6 +3524,7 @@ class ExprAvailabilityWalker : public ASTWalker {
   ASTContext &Context;
   MemberAccessContext AccessContext = MemberAccessContext::Default;
   SmallVector<const Expr *, 16> ExprStack;
+  SmallVector<bool, 4> PreconcurrencyCalleeStack;
   const ExportContext &Where;
 
 public:
@@ -3546,6 +3547,15 @@ public:
     auto *DC = Where.getDeclContext();
 
     ExprStack.push_back(E);
+
+    if (auto *apply = dyn_cast<ApplyExpr>(E)) {
+      bool preconcurrency = false;
+      auto declRef = apply->getFn()->getReferencedDecl();
+      if (auto *decl = declRef.getDecl()) {
+        preconcurrency = decl->preconcurrency();
+      }
+      PreconcurrencyCalleeStack.push_back(preconcurrency);
+    }
 
     if (auto DR = dyn_cast<DeclRefExpr>(E)) {
       diagnoseDeclRefAvailability(DR->getDeclRef(), DR->getSourceRange(),
@@ -3669,9 +3679,15 @@ public:
                                               EE->getLoc(),
                                               Where.getDeclContext());
 
+      bool preconcurrency = false;
+      if (!PreconcurrencyCalleeStack.empty()) {
+        preconcurrency = PreconcurrencyCalleeStack.back();
+      }
+
       for (ProtocolConformanceRef C : EE->getConformances()) {
         diagnoseConformanceAvailability(E->getLoc(), C, Where, Type(), Type(),
-                                        /*useConformanceAvailabilityErrorsOpt=*/true);
+                                        /*useConformanceAvailabilityErrorsOpt=*/true,
+                                        /*preconcurrency=*/preconcurrency);
       }
     }
 
@@ -3696,6 +3712,10 @@ public:
   PostWalkResult<Expr *> walkToExprPost(Expr *E) override {
     assert(ExprStack.back() == E);
     ExprStack.pop_back();
+
+    if (auto *apply = dyn_cast<ApplyExpr>(E)) {
+      PreconcurrencyCalleeStack.pop_back();
+    }
 
     return Action::Continue(E);
   }

--- a/lib/Sema/TypeCheckAvailability.cpp
+++ b/lib/Sema/TypeCheckAvailability.cpp
@@ -53,7 +53,8 @@ concreteSyntaxDeclForAvailableAttribute(const Decl *AbstractSyntaxDecl);
 static bool diagnoseExplicitUnavailability(
     SourceLoc loc, const RootProtocolConformance *rootConf,
     const ExtensionDecl *ext, const ExportContext &where,
-    bool warnIfConformanceUnavailablePreSwift6 = false);
+    bool warnIfConformanceUnavailablePreSwift6 = false,
+    bool preconcurrency = false);
 
 /// Emit a diagnostic for references to declarations that have been
 /// marked as unavailable, either through "unavailable" or "obsoleted:".
@@ -2986,7 +2987,8 @@ getExplicitUnavailabilityDiagnosticInfo(const Decl *decl,
 bool diagnoseExplicitUnavailability(
     SourceLoc loc, const RootProtocolConformance *rootConf,
     const ExtensionDecl *ext, const ExportContext &where,
-    bool warnIfConformanceUnavailablePreSwift6) {
+    bool warnIfConformanceUnavailablePreSwift6,
+    bool preconcurrency) {
   // Invertible protocols are never unavailable.
   if (rootConf->getProtocol()->getInvertibleProtocolKind())
     return false;
@@ -3012,7 +3014,7 @@ bool diagnoseExplicitUnavailability(
   diags
       .diagnose(loc, diag::conformance_availability_unavailable, type, proto,
                 platform.empty(), platform, EncodedMessage.Message)
-      .limitBehaviorUntilSwiftVersion(behavior, 6)
+      .limitBehaviorWithPreconcurrency(behavior, preconcurrency)
       .warnUntilSwiftVersionIf(warnIfConformanceUnavailablePreSwift6, 6);
 
   switch (diagnosticInfo->getStatus()) {
@@ -4641,7 +4643,8 @@ swift::diagnoseConformanceAvailability(SourceLoc loc,
       // FIXME: diagnoseExplicitUnavailability() should take unmet requirement
       if (diagnoseExplicitUnavailability(
               loc, rootConf, ext, where,
-              warnIfConformanceUnavailablePreSwift6)) {
+              warnIfConformanceUnavailablePreSwift6,
+              preconcurrency)) {
         maybeEmitAssociatedTypeNote();
         return true;
       }

--- a/test/Concurrency/predates_concurrency_swift6.swift
+++ b/test/Concurrency/predates_concurrency_swift6.swift
@@ -198,7 +198,7 @@ func requireSendable<T: Sendable>(_: T) {}
 @preconcurrency
 struct RequireSendable<T: Sendable> {}
 
-class NotSendable {} // expected-note 2 {{class 'NotSendable' does not conform to the 'Sendable' protocol}}
+class NotSendable {} // expected-note 4 {{class 'NotSendable' does not conform to the 'Sendable' protocol}}
 
 typealias T = RequireSendable<NotSendable>
 // expected-warning@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
@@ -206,4 +206,25 @@ typealias T = RequireSendable<NotSendable>
 func testRequirementDowngrade(ns: NotSendable) {
   requireSendable(ns)
   // expected-warning@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
+}
+
+
+protocol P2 {}
+
+extension NotSendable: P2 {}
+
+@preconcurrency
+func requireSendableExistential(_: any P2 & Sendable) {}
+
+func requireSendableExistentialAlways(_: any P2 & Sendable) {}
+
+func testErasureDowngrade(ns: NotSendable) {
+  requireSendableExistential(ns)
+  // expected-warning@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
+
+  withSendableClosure {
+    let ns = NotSendable()
+    requireSendableExistentialAlways(ns)
+    // expected-error@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
+  }
 }

--- a/test/Concurrency/predates_concurrency_swift6.swift
+++ b/test/Concurrency/predates_concurrency_swift6.swift
@@ -200,12 +200,24 @@ struct RequireSendable<T: Sendable> {}
 
 class NotSendable {} // expected-note 4 {{class 'NotSendable' does not conform to the 'Sendable' protocol}}
 
+class UnavailableSendable {}
+
+@available(*, unavailable)
+extension UnavailableSendable: @unchecked Sendable {}
+// expected-note@-1 4 {{conformance of 'UnavailableSendable' to 'Sendable' has been explicitly marked unavailable here}}
+
 typealias T = RequireSendable<NotSendable>
 // expected-warning@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
 
-func testRequirementDowngrade(ns: NotSendable) {
+typealias T2 = RequireSendable<UnavailableSendable>
+// expected-warning@-1 {{conformance of 'UnavailableSendable' to 'Sendable' is unavailable}}
+
+func testRequirementDowngrade(ns: NotSendable, us: UnavailableSendable) {
   requireSendable(ns)
   // expected-warning@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
+
+  requireSendable(us)
+  // expected-warning@-1 {{conformance of 'UnavailableSendable' to 'Sendable' is unavailable}}
 }
 
 
@@ -213,18 +225,27 @@ protocol P2 {}
 
 extension NotSendable: P2 {}
 
+extension UnavailableSendable: P2 {}
+
 @preconcurrency
 func requireSendableExistential(_: any P2 & Sendable) {}
 
 func requireSendableExistentialAlways(_: any P2 & Sendable) {}
 
-func testErasureDowngrade(ns: NotSendable) {
+func testErasureDowngrade(ns: NotSendable, us: UnavailableSendable) {
   requireSendableExistential(ns)
   // expected-warning@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
+
+  requireSendableExistential(us)
+  // expected-warning@-1 {{conformance of 'UnavailableSendable' to 'Sendable' is unavailable}}
 
   withSendableClosure {
     let ns = NotSendable()
     requireSendableExistentialAlways(ns)
     // expected-error@-1 {{type 'NotSendable' does not conform to the 'Sendable' protocol}}
+
+    let us = UnavailableSendable()
+    requireSendableExistentialAlways(us)
+    // expected-error@-1 {{conformance of 'UnavailableSendable' to 'Sendable' is unavailable}}
   }
 }

